lib/HaskellLib_H.thy

```diff
@@ -548,6 +548,30 @@ lemma gets_the_ohaskell_assert:
   "gets_the (ohaskell_assert P []) = assert P"
   by (clarsimp simp: ohaskell_assert_def split: if_splits)
 
-lemmas omonad_defs = omonad_defs ohaskell_assert_def ohaskell_fail_def
+definition ostate_assert :: "('s \<Rightarrow> bool) \<Rightarrow> ('s, unit) lookup" where
+  "ostate_assert P \<equiv> do { s \<leftarrow> ask; oassert (P s)}"
+
+definition ohaskell_state_assert :: "('s \<Rightarrow> bool) \<Rightarrow> unit list \<Rightarrow> ('s, unit) lookup" where
+  "ohaskell_state_assert P L \<equiv> ostate_assert P"
+
+lemma gets_the_ostate_assert:
+  "gets_the (ostate_assert P) = stateAssert P []"
+  by (clarsimp simp: ostate_assert_def stateAssert_def state_assert_def gets_the_Some_get)
```

Comment:
This should not use the Haskell `stateAssert` on the rhs, but the library `state_assert` instead and should move to the Monad library.

(We should also have a rewrite from `stateAssert` to `state_assert` somewhere so that we can get this equality automatically if we need it).

Reply:
The annoying thing here is that so many of our `corres/ccorres` rules are stated in terms of `stateAssert`, so if we wanted to use them with `state_assert`, we would currently have to do a whole bunch of `[simplified stateAssert_def]`ing. Would it be worth making any changes to prefer `state_assert`?

Reply:
Right, I see where this is coming from. We could add a rule that does `state_assert P = stateAssert P []` and use that with simp before we apply `corres` rules, but that does feel a bit dirty.

I'd prefer having a global `simp` that turns `stateAssert` into `state_assert` and changing the corres rules, but that could potentially be quite a few of them. Or, maybe less invasive: have variants of the stateAssert corres rules (via `lemmas`, that already are in the `state_assert` form). Then at least the application side is fine and can be adapted incrementally.

Reply:
Should I change this to `state_assert`, and make an issue for writing all of our `corres/ccorres` rules in terms of `state_assert`?

Reply:
Issue is good. I think I'd prefer having both variants for the `corres` rules, because sometimes you don't want to run `simp` first.

Reply:
OK I will write up an issue. Or seeing that we never use the string once it's in the design spec, are we really all that interested in using `stateAssert`? Could we get rid of `stateAssert` and just use `state_assert`? I think we could just change the Haskell translator in one or two spots to achieve this, maybe

Reply:
I was thinking about the same thing last night. Yes, we should look at what it would take to make the Haskell translator produce `state_assert` directly. Maybe put both options into the issue. It's possible that the translator change will lead to relatively big changes in the proofs.

Reply:
Issue here: https://github.com/seL4/l4v/issues/787